### PR TITLE
compute `inline_worthy` after inference and cache it

### PR DESCRIPTION
**Pull request by JeffBezanson**
This avoids repeating the computation for the same piece of code, and allows rejecting inlining before calling jl_uncompress_ast in more cases.

This should be generally faster. The only reason it might be slower AFAICT would be if we end up calling inline_worthy for lots of functions that previously never had inlining attempts by call sites. However inline_worthy itself should be pretty cheap.

### CODE REVIEW

**Comment by Keno** [comment]
Does it make sense to instead compute this the first time we ask for it?

**Comment by JeffBezanson** [comment]
Good idea; I'll try that. I guess the main annoyance is that there would be 3 states, which is awkward to represent.

File: base/inference.jl [hunk 1]
```diff
@@ -1877,6 +1877,26 @@ function finish(me::InferenceState)
...
     me.inferred = true
 
+    # determine and cache inlineability
+    inlineable = false
+    if isdefined(me.linfo, :def)
+        cost = 1000
+        if me.linfo.def.module === _topmod(me)
+            name = me.linfo.def.name
+            if me.linfo.isva && (name === :+ || name === :* || name === :min || name === :max)
+                inlineable = true
```
Comment (vtjnash) [inline base/inference.jl]: maybe just `cost ÷= 20`? i think the previous test seemed very strict on which method it was matching to hopefully avoid the corner cases where inlining would make it slower.

**Comment by vtjnash** [comment]
i don't think it's going to be worth deferring the test. if we have the information at the read like this, i believe the system should be able to delete this code as soon as it is converted into native code, thus saving memory.

**Comment by JeffBezanson** [comment]
:+1: to that

**Comment by JeffBezanson** [comment]
Ok, I have implemented that. It definitely seems to save significant memory during some tests. However it only shaves 1-2MB off the system image, which was disappointing. My best guess is that this is due to having copies of LambdaInfo in both method caches and tfunc caches. #15918 might help towards that (plus we also need to eliminate `InferenceState.destination`).

The most troublesome point I ran into was the serialize test. Method.roots contained references to codegen'd LambdaInfos whose IR had been deleted. AFAICT, there should be no need to serialize codegen'd LambdaInfos. All of these roots came from the recently-added `jl_add_linfo_root(ctx->linfo, (jl_value_t*)li);` in `emit_call`. I don't understand why this root would be necessary --- generated code doesn't reference its LambdaInfo object, and any referenced values should have been added to some Method.roots array anyway.

**Comment by vtjnash** [comment]
Iirc, that link keeps the method alive if it gets replaced. I'm thinking for #265 that method replacement will just mark them rather thnq deleting them, avoiding that problem

**Comment by JeffBezanson** [comment]
I see; so indeed that specific LambdaInfo is not needed, but probably the called function's def.roots. I'll try changing it to root li.def.roots instead.

**Comment by JeffBezanson** [comment]
Here we have an OSX timeout and a crash on 32-bit AV. The crash is a bit worrying.

File: src/codegen.cpp [hunk 2]
```diff
@@ -2842,8 +2858,10 @@ static jl_cgval_t emit_call(jl_expr_t *ex, jl_codectx_t *ctx)
                 else {
                     fval = emit_expr(args[0], ctx);
                 }
-                if (ctx->linfo->def) // root this li in case it gets deleted from the cache in `f`
-                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li);
+                // keep a reference to the called function's roots array in case
+                // the method is replaced (PR #14301)
+                if (ctx->linfo->def && li->def && li->def->roots)
+                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li->def->roots);
```
Comment (vtjnash) [inline src/codegen.cpp]: another possible source of roots is `sparam_vals`, which `emit_sparam` assumes has a root (but could be any isbits)

additionally, lets just stop removing methods from the cache once they've been returned from `jl_get_specialization1`. this should be a really small number (since we actively discourage overwriting methods), avoids this issue, and is a going to be needed for 265 anyways.

Comment (JeffBezanson) [inline src/codegen.cpp]: Yeah, that sounds reasonable, and much less of a hack than adding roots in callers. How do we reference the old entry in the cache? An `old_entry` link in TypeMapEntry?

Comment (vtjnash) [inline src/codegen.cpp]: I was thinking this would be a decent opportunity to incrementally start to add the entry world counters (min / max). an entry would be invisible if the world counter is not between min (this would be monotonically incrementing on every method add and copied from every method definition to its cache entries) and max (initialized to typemax). initially, the TLS world counter can just be equal to the method add counter.

Comment (vtjnash) [inline src/codegen.cpp]: maybe just take this out of the diff (it doesn't seem related to the rest of the change), and then it can be merged?

Comment (JeffBezanson) [inline src/codegen.cpp]: Ok sounds good.

**Comment by JeffBezanson** [comment]
Closing but will keep this branch around for the second commit.
